### ISSUE TITLE
ssr: on apollo error the treewalker has no catch and breaks whole app

### ISSUE THREAD

**Issue by maapteh** [apollographql/react-apollo#3918]
**Intended outcome:**
When having an Apollo Error thrown via GraphQL, the React ssr treewalker should just take the Error and proceed.

**Actual outcome:**
When we have one query with an Apollo Error, which get sent in batch to server, then the treewalker just [throws the Error!](https://github.com/apollographql/react-apollo/blob/master/packages/ssr/src/getDataFromTree.ts#L53) and breaks your application from rendering at all.

```
[web] (node:90112) UnhandledPromiseRejectionWarning: Error: GraphQL error: member oepsie
[web]     at new ApolloError (SNIP/node_modules/apollo-client/bundle.umd.js:92:26)
```

For client-side eveything works as it should. The errors are passed to error object, data is null.

**How to reproduce the issue:**
Have two queries, let one of them sent directly an Error, do both React query components ssr using treewalker.

**Version**
```
"@apollo/react-common": "3.1.3",
"@apollo/react-hooks": "3.1.3",
"@apollo/react-ssr": "3.1.3",
"@apollo/react-testing": "3.1.3",

"apollo-cache-inmemory": "1.6.5",
"apollo-client": "2.6.8",
"apollo-link-batch-http": "1.2.13",
"apollo-link-error": "1.1.12",
"apollo-link-http": "1.5.16",
"graphql": "14.6.0",
"graphql-tag": "2.10.3",
```
*Issue was related to #1403, but that one got closed and seems to not getting any response anymore so opening new one.*

**@hwillson if you need reproduction, i can create it**

**Comment by maapteh** [comment]
Wow, so setting `await getDataFromTree(appTree, { errorPolicy: 'all' });` solves it, but i dont understand, who wants their app to break? Why is the default to let the treewalker just throw the error?

**Comment by maapteh** [comment]
Having errorPolicy in the query which throws fixes it, then it will not throw the error. But having the errorPolicy as default in the apollo client doesn't :(

so good:
```
    const { data, error } = useMemberAddressQuery({
        ssr: true,
        errorPolicy: 'all',
    });
```

bad:
```
    const client = new ApolloClient({
        cache,
        resolvers: {},
        ...SNIP
        queryDeduplication: true,
        defaultOptions: {
            mutate: {
                errorPolicy: 'all',
            },
            query: {
                errorPolicy: 'all',
            },
        },
    });
```
Howcome was errorPolicy not the default?

**Comment by maapteh** [comment]
Shit it was in my generated hooks :( Sorry Team. For others, if you set the default options it will work. It will not throw the error. I dont know who in hell ever wants his app treewalker to literally throw an error but hey :) So this will fix your issues:

```
    const client = new ApolloClient({
        cache,
        resolvers: {},
        ...SNIP
        queryDeduplication: true,
        defaultOptions: {
            mutate: {
                errorPolicy: 'all',
            },
            query: {
                errorPolicy: 'all',
            },
        },
    });
```

**Comment by maapteh** [comment]
Ah damn, with that policy being set i loose my error itself. So re-opening. I cant find the correct settings when throwing error on SSR.

- when no error policy set it will throw the error while tree walking and make the whole app broken
- when error policy set to all it will not throw the error, but the error is not being passed to the component, so data is null but the whole reason why we used to throw the error on the graphql part is to use the error to show user technical error etc etc

Im lost...

**Comment by maapteh** [comment]
when using 3.1.5 then error is not passed :( 3.1.3 i het error from client query... im super lost

**Comment by michaelchiche** [comment]
@maapteh you should create a codesandbox repro

**Comment by maapteh** [comment]
Yes maybe i will, point is then that i have to spin up the api and web and hugh already has my sample app which is more easier to share :)

**Comment by maapteh** [comment]
Sandbox created, will also update the issue accordingly:

https://codesandbox.io/s/github/maapteh/sandbox-react-apollo-issues-3918

So this app shows how to prevent the error being thrown on treewalker but also that there is no error for the component to handle :( When you remove errorPolicy then treewalker throws (which we catch there) but then the call is also done client-side and then fills the error. But that puts load on my server twice. I just want to throw the error once :)

**Comment by maapteh** [comment]
Im now finding more and more issues related to this bug, like https://github.com/apollographql/apollo-client/issues/4644 

When i started looking in the react package i only found this closed one. Hopefully the sandbox helps in identifying the rootcause and close these issues and make lots of users happy :)

**Comment by maapteh** [comment]
@hwillson if i would know how to solve it, i would have made a PR. But me and my team is super lost in this, and it also seems to be a bug for a long time (went to latest alpha and back to 2). Is it possible to give some hints, or even better fix it together?
